tests/unit_tests/test_matchmaker_algorithm_bucket_teams.py

```diff
@@ -0,0 +1,312 @@
+        ladder_games: int = config.NEWBIE_MIN_GAMES + 1,
+        name=None,
+    ):
+        """Make a player with the given ratings"""
+        player = player_factory(
+            ladder_rating=(mean, deviation),
+            ladder_games=ladder_games,
+            login=name,
+            lobby_connection_spec=None,
+        )
+        return player
+
+    return make
+
+
+@pytest.mark.parametrize("make_teams_func", (_make_teams, _make_teams_from_single))
+@given(
+    searches=st_searches_list(max_players=1),
+    size=st.integers(min_value=1, max_value=10),
+)
+def test_make_teams_single_correct_size(searches, size, make_teams_func):
+    matched, _ = make_teams_func(searches, size)
+
+    for search in matched:
+        assert len(search.players) == size
+
+
+def test_make_teams_single_2v2_large_pool(player_factory):
+    """
+    When we have a large number of players all with similar ratings, we want
+    teams to be formed by putting players with the same rating on the same team.
+    """
+
+    # Large enough so the test is unlikely to pass by chance
+    num = 40
+
+    searches = [
+        Search([player_factory(random.uniform(950, 1050), 10, name=f"p{i}")])
+        for i in range(num)
+    ]
+    searches += [
+        Search([player_factory(random.uniform(450, 550), 10, name=f"p{i}")])
+        for i in range(num)
+    ]
+    matched, non_matched = _make_teams_from_single(searches, size=2)
+
+    assert matched != []
+    assert non_matched == []
+
+    for search in matched:
+        p1, p2 = search.players
+        p1_mean, _ = p1.ratings[RatingType.LADDER_1V1]
+        p2_mean, _ = p2.ratings[RatingType.LADDER_1V1]
+        #
+        assert math.fabs(p1_mean - p2_mean) <= 100
+
+
+def test_make_teams_single_2v2_small_pool(player_factory):
+    """
+    When we have a small number of players, we want teams to be formed by
+    distributing players of equal skill to different teams so that we can
+    maximize the chances of getting a match.
+    """
+
+    # Try a bunch of times so it is unlikely to pass by chance
+    for _ in range(20):
+        searches = [
+            Search([player_factory(random.gauss(1000, 5), 10, name=f"p{i}")])
+            for i in range(2)
+        ]
+        searches += [
+            Search([player_factory(random.gauss(500, 5), 10, name=f"r{i}")])
+            for i in range(2)
+        ]
+        matched, non_matched = _make_teams_from_single(searches, size=2)
+
+        assert matched != []
+        assert non_matched == []
+
+        for search in matched:
+            p1, p2 = search.players
+            # Order doesn't matter
+            if p1.ratings[RatingType.LADDER_1V1][0] > 900:
+                assert p2.ratings[RatingType.LADDER_1V1][0] < 600
+            else:
+                assert p1.ratings[RatingType.LADDER_1V1][0] < 600
+                assert p2.ratings[RatingType.LADDER_1V1][0] > 900
+
+
+def test_make_buckets_performance(bench, player_factory):
```

Comment:
Isn't there a pytest mark for bench tests now? Or is that added by @BlackYps?

Reply:
Yeah, that gets added by my PR